experimental/oak_baremetal_channel/src/lib.rs

```diff
@@ -350,28 +50,70 @@ where
 {
     pub fn new(socket: T) -> Self {
         Self {
-            inner: Framed::new(socket),
+            inner: frame::Framed::new(socket),
         }
     }
-    fn recursively_complete_partial_message(
-        &mut self,
-        partial_message: PartialMessage,
-    ) -> anyhow::Result<Message> {
-        let frame = self.inner.read_frame()?;
-        match partial_message.add_frame(frame)? {
-            CompletionResult::Complete(message) => Ok(message),
-            CompletionResult::Incomplete(partial_message) => {
-                self.recursively_complete_partial_message(partial_message)
+    pub fn read_message<M: message::Message>(&mut self) -> anyhow::Result<M> {
+        let mut encoded_message: Vec<u8> = Vec::new();
+        let mut first_frame = self.inner.read_frame().context("failed to read frame.")?;
+
+        match (
+            first_frame.flags.contains(frame::Flags::START),
+            first_frame.flags.contains(frame::Flags::END),
+        ) {
+            (true, true) => {
+                return Ok(M::decode(first_frame.body));
+            }
+            (true, false) => {
+                let required_additional_capacity_to_hold_message = {
+                    let message_length: usize = {
+                        let mut message_length_bytes: [u8; message::LENGTH_SIZE] =
+                            [0; message::LENGTH_SIZE];
+                        let message_length_offset = frame::BODY_OFFSET + message::LENGTH_OFFSET;
+                        let message_length_range =
+                            message_length_offset..(message_length_offset + message::LENGTH_SIZE);
+                        message_length_bytes
+                            .copy_from_slice(&first_frame.body[message_length_range]);
+                        usize::try_from(message::Length::from_le_bytes(message_length_bytes))
+                            .expect("couldn't convert message lemgth to usize")
+                    };
+                    message_length - encoded_message.capacity()
+                };
+                if required_additional_capacity_to_hold_message > 0 {
+                    encoded_message.reserve_exact(required_additional_capacity_to_hold_message);
+                }
+
+                encoded_message.append(&mut first_frame.body);
+            }
+            _ => {
+                anyhow::bail!("expected a frame with the START flag set.")
             }
+        };
+
+        loop {
+            let mut frame = self.inner.read_frame().context("failed to read frame.")?;
+
+            match frame.flags.contains(frame::Flags::START) {
```

Comment:
nit: once again personal prefernce, but I find it easier to read if statements rather than a match on a bool.

Reply:
+1